[SPARK-26718][SS] Fixed integer overflow in SS kafka rateLimit calculation

linehrr opened the pull request:
## What changes were proposed in this pull request?

Fix the integer overflow issue in rateLimit.

## How was this patch tested?

Pass the Jenkins with newly added UT for the possible case where integer could be overflowed.

Commit:
fixed integer overflow in rateLimit calculation

added UT to against the calculation, setting maxRatePerTrigger to Integer.MaxValue
 Changes to be committed:
	modified:   src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala
	modified:   src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala
	modified:   src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

dongjoon-hyun commented:
ok to test

SparkQA commented:
**[Test build #101744 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101744/testReport)** for PR 23666 at commit [`03fe7f9`](https://github.com/apache/spark/commit/03fe7f93549af9856d08b5471893837898b7242f).

dongjoon-hyun commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:
```diff
@@ -204,6 +204,40 @@ abstract class KafkaMicroBatchSourceSuiteBase extends KafkaSourceSuiteBase {
+    val topic = newTopic()
+    testUtils.createTopic(topic, partitions = 1)
+    testUtils.sendMessages(topic, (0 until 5000).map(_.toString).toArray, Some(0))   // fill in 5000 messages to trigger potential integer overflow
+
```
Shall we remove empty line 211 since we have two empty line (211~212)?

SparkQA commented:
**[Test build #101744 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101744/testReport)** for PR 23666 at commit [`03fe7f9`](https://github.com/apache/spark/commit/03fe7f93549af9856d08b5471893837898b7242f).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101744/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

HeartSaVioR commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala:
```diff
@@ -231,7 +231,11 @@ private[kafka010] class KafkaMicroBatchStream(
+            // need to be careful of integer overflow
+            // therefore added canary checks where to see if off variable could be overflowed
+            // refer to [https://issues.apache.org/jira/browse/SPARK-26718]
+            val off = if (prorateLong > Long.MaxValue - begin) Long.MaxValue else begin + prorateLong
```
Line length is limited to 100 chars and this line exceeds it.
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101744/console

You are encouraged to run `dev/scalastyle` and fix all style issues before submitting additional commit.

HeartSaVioR commented:
You might be confused regarding policy for the first time, but we only add `[BRANCH-NAME]` for PRs against non-master branch. You can remove `[Master]` in PR's title.

Commit:
linting code

 On branch master
 Your branch is up to date with 'origin/master'.

 Changes to be committed:
	modified:   external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala
	modified:   external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala
	modified:   external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7490/
Test PASSed.

SparkQA commented:
**[Test build #101766 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101766/testReport)** for PR 23666 at commit [`06ddee2`](https://github.com/apache/spark/commit/06ddee277a21b8cf1461e132513e55f492aad341).

SparkQA commented:
**[Test build #101766 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101766/testReport)** for PR 23666 at commit [`06ddee2`](https://github.com/apache/spark/commit/06ddee277a21b8cf1461e132513e55f492aad341).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101766/
Test PASSed.

arunmahadevan reviewed (approved):
Looks good. Would be good to back port this to older versions (Spark 2.4, 2.3 and so on)

HeartSaVioR reviewed (approved):
LGTM

dongjoon-hyun commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:
```diff
@@ -204,6 +204,43 @@ abstract class KafkaMicroBatchSourceSuiteBase extends KafkaSourceSuiteBase {
   }
 
+  test("Rate limit set to Long.Max should not overflow integer " +
+    "during end offset calculation[SPARK-26718]") {
```
Hi, @linehrr .
We usually put SPARK ID at the beginning of the test.
```scala
  test("SPARK-26718 Rate limit set to Long.Max should not overflow integer " +
    "during end offset calculation") {
```

linehrr replied:
ok I got the idea that ticket number should be at the end from test case: `  test("ensure that initial offset are written with an extra byte in the beginning (SPARK-19517)") {`, but I can put it at the front.

Commit:
adjust ticket number position in test desc

 Changes to be committed:
	modified:   src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala

SparkQA commented:
**[Test build #101769 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101769/testReport)** for PR 23666 at commit [`7a5da0f`](https://github.com/apache/spark/commit/7a5da0f982859e0c03f04014ca696b641c51a682).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7491/
Test PASSed.

dongjoon-hyun commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:
```diff
@@ -204,6 +204,43 @@ abstract class KafkaMicroBatchSourceSuiteBase extends KafkaSourceSuiteBase {
+      // use latest to force begin to be 5000
+      .option("startingOffsets", startingOffsets)
+      // use Long.Max to try to trigger overflow
+      .option("maxOffsetsPerTrigger", Long.MaxValue.toString)
```
nit. We can use
```scala
- .option("maxOffsetsPerTrigger", Long.MaxValue.toString)
+ .option("maxOffsetsPerTrigger", Long.MaxValue)
```

linehrr replied:
oh, good point. actually wish SparkConf.set() has such overloads as well, LOL.

Commit:
removed unnecessary type conversion

 Changes to be committed:
	modified:   src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala

SparkQA commented:
**[Test build #101771 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101771/testReport)** for PR 23666 at commit [`48509a5`](https://github.com/apache/spark/commit/48509a5a69612a013b4844859a8c401790788868).

dongjoon-hyun commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:
```diff
@@ -204,6 +204,43 @@ abstract class KafkaMicroBatchSourceSuiteBase extends KafkaSourceSuiteBase {
+    val startingOffsets = JsonUtils.partitionOffsets(partitionOffsets)
+
+    val sparkSession = spark
+    import sparkSession.implicits._
```
Are the above two lines required? Shall we remove these?
```scala
-    val sparkSession = spark
-    import sparkSession.implicits._
```

linehrr replied:
ha, it's not, Encoders are already provided in this context as implicits, I will remove it.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7492/
Test PASSed.

SparkQA commented:
**[Test build #101769 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101769/testReport)** for PR 23666 at commit [`7a5da0f`](https://github.com/apache/spark/commit/7a5da0f982859e0c03f04014ca696b641c51a682).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101769/
Test PASSed.

dongjoon-hyun commented:
@linehrr. I updated the PR description a little.

Commit:
removed useless implicits imports

 Changes to be committed:
	modified:   src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala

SparkQA commented:
**[Test build #101771 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101771/testReport)** for PR 23666 at commit [`48509a5`](https://github.com/apache/spark/commit/48509a5a69612a013b4844859a8c401790788868).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101771/
Test PASSed.

dongjoon-hyun commented:
@linehrr . The PR itself and the test case looks correct. The last concern I have is the UT overhead. This single test case costs us almost 3 minutes since it works like a kind of end-to-end integration test. It's a little huge since your test case will survive for a long time. I'm wondering if you can minimize the overhead while achieving the same test coverage?

```
[info] - SPARK-26718 Rate limit set to Long.Max should not overflow integer during end offset calculation (2 minutes, 57 seconds)
```

SparkQA commented:
**[Test build #101772 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101772/testReport)** for PR 23666 at commit [`30f48e8`](https://github.com/apache/spark/commit/30f48e8bc1c8ea76a891a103b511f7efc7ec3fed).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101772/
Test PASSed.

linehrr commented:
> I'm wondering if you can minimize the overhead while achieving the same test coverage?

@dongjoon-hyun, I think I could. let me look into that. as long as I can prove test case will be able to trigger old bug then it should be a sufficient test.

dongjoon-hyun commented:
Thank you a lot, @linehrr !

Commit:
made test case faster

 Changes to be committed:
	modified:   src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala

SparkQA commented:
**[Test build #101834 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101834/testReport)** for PR 23666 at commit [`1700c34`](https://github.com/apache/spark/commit/1700c3428a651f6d35174bd73b5f2029c26432bc).

Commit:
fixed comments

 Changes to be committed:
	modified:   src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7538/
Test PASSed.

linehrr commented:
@dongjoon-hyun after some quick investigation, I found out that it's 
`    testUtils.sendMessages(topic, (0 until 5000).map(_.toString).toArray, Some(0))`
that's slowing down the whole test case, it can sometimes take up to 6 seconds to generate those 5000 messages(kinda make sense). I put this 5000 magic number there to make sure it can trigger the overflow due to `begin + prorateLong`. but then I also found out that merely 5 messages could achieve the same purpose with no problem. and this brought the message generation time from 5-6 seconds to only about 0.1 second(make sense with only 0.1% volume compare to before). 
based on above, I changed the number from 5000 to 5. and also verified that it triggers error from both V1Datasource and V2Datasource. therefore I consider it as a valid test case still.

SparkQA commented:
**[Test build #101835 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101835/testReport)** for PR 23666 at commit [`9a12501`](https://github.com/apache/spark/commit/9a125017d79887e146977ed6ae153fd833453937).

SparkQA commented:
**[Test build #101834 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101834/testReport)** for PR 23666 at commit [`1700c34`](https://github.com/apache/spark/commit/1700c3428a651f6d35174bd73b5f2029c26432bc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101834/
Test PASSed.

SparkQA commented:
**[Test build #101835 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101835/testReport)** for PR 23666 at commit [`9a12501`](https://github.com/apache/spark/commit/9a125017d79887e146977ed6ae153fd833453937).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101835/
Test PASSed.

dongjoon-hyun reviewed (approved):
+1, LGTM. Merged to master.

Thank you for your first contribution, @linehrr ! :D
